File: core/src/main/java/com/stevenlagoy/presidency/map/travel/route/Seaport.java
package com.stevenlagoy.presidency.map.travel.route;

import com.stevenlagoy.jsonic.JSONObject;
import com.stevenlagoy.jsonic.Jsonic;
import com.stevenlagoy.presidency.map.Municipality;

public class Seaport extends Route implements Jsonic<Seaport> {

    private String fullName;
    private String commonName;
    private Municipality location;

    public Seaport() {
        fullName = "";
        commonName = "";
        location = null;
    }

    public Seaport(String fullName, String commonName, Municipality location) {
        this.fullName = fullName;
        this.commonName = commonName;
        this.location = location;
    }

    public String getFullName() {
        return fullName;
    }

    public void setFullName(String name) {
        this.fullName = name;
    }

    public String getCommonName() {
        return commonName;
    }

    public void setCommonName(String name) {
        this.commonName = name;
    }

    public Municipality getMunicipality() {
        return location;
    }

    public void setMunicipality(Municipality location) {
        this.location = location;
    }

    // REPRESENTATION METHODS
    // ---------------------------------------------------------------------

    @Override
    public JSONObject toJson() {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'toJson'");
    }

    @Override
    public Seaport fromJson(JSONObject json) {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'fromJson'");
    }
}

File: core/src/main/kotlin/com/stevenlagoy/presidency/map/travel/route/Seaport.kt
package com.stevenlagoy.presidency.map.travel.route

import com.stevenlagoy.jsonic.JSONObject
import com.stevenlagoy.presidency.core.Engine
import com.stevenlagoy.presidency.map.Municipality

class Seaport(
    engine: Engine,
    name: String = "",
    location: Municipality? = null,
    connections: List<Municipality> = listOf(),
) : Route(engine, name, connections) {

    var location: Municipality? = location
        internal set

    constructor(engine: Engine, json: JSONObject) : this (engine) {
        fromJson(json)
    }

    override fun toJson() = JSONObject(name, listOf(
        JSONObject("name", name),
    ))

    override fun fromJson(json: JSONObject) = this.apply {
        super.fromJson(json)
        location = engine.MAP_MANAGER.matchMunicipality(json.requireString("location")).orElseThrow { IllegalArgumentException("Could not find location for seaport $name") }
    }

}
